src/search.rs

```diff
@@ -1,202 +1,261 @@
+        self.offset = Some(offset);
+        self
     }
-    pub fn with_facet_filters(self, facet_filters: Vec<Vec<&'a str>>) -> Query<'a> {
-        Query {
-            facet_filters: Some(facet_filters),
-            ..self
-        }
+    pub fn with_limit<'b>(&'b mut self, limit: usize) -> &'b mut Query<'a> {
+        self.limit = Some(limit);
+        self
     }
-    pub fn with_facets_distribution(self, facets_distribution: Option<Vec<&'a str>>) -> Query<'a> {
-        Query {
-            facets_distribution: Some(facets_distribution),
-            ..self
-        }
+    pub fn with_filters<'b>(&'b mut self, filters: &'a str) -> &'b mut Query<'a> {
+        self.filters = Some(filters);
+        self
     }
-    pub fn with_attributes_to_retrieve(self, attributes_to_retrieve: &'a str) -> Query<'a> {
-        Query {
-            attributes_to_retrieve: Some(attributes_to_retrieve),
-            ..self
-        }
+    pub fn with_facet_filters<'b>(&'b mut self, facet_filters: &'a[&'a[&'a str]]) -> &'b mut Query<'a> {
+        self.facet_filters = Some(facet_filters);
+        self
     }
-    pub fn with_attributes_to_crop(self, attributes_to_crop: &'a str) -> Query<'a> {
-        Query {
-            attributes_to_crop: Some(attributes_to_crop),
-            ..self
-        }
+    pub fn with_facets_distribution<'b>(&'b mut self, facets_distribution: Selectors<&'a[&'a str]>) -> &'b mut Query<'a> {
+        self.facets_distribution = Some(facets_distribution);
+        self
     }
-    pub fn with_attributes_to_highlight(self, attributes_to_highlight: &'a str) -> Query<'a> {
-        Query {
-            attributes_to_highlight: Some(attributes_to_highlight),
-            ..self
-        }
+    pub fn with_attributes_to_retrieve<'b>(&'b mut self, attributes_to_retrieve: Selectors<&'a [&'a str]>) -> &'b mut Query<'a> {
+        self.attributes_to_retrieve = Some(attributes_to_retrieve);
+        self
     }
-    pub fn with_crop_length(self, crop_length: usize) -> Query<'a> {
-        Query {
-            crop_length: Some(crop_length),
-            ..self
-        }
+    pub fn with_attributes_to_crop<'b>(&'b mut self, attributes_to_crop: Selectors<&'a [(&'a str, Option<usize>)]>) -> &'b mut Query<'a> {
+        self.attributes_to_crop = Some(attributes_to_crop);
+        self
+    }
+    pub fn with_attributes_to_highlight<'b>(&'b mut self, attributes_to_highlight: Selectors<&'a [&'a str]>) -> &'b mut Query<'a> {
+        self.attributes_to_highlight = Some(attributes_to_highlight);
+        self
+    }
+    pub fn with_crop_length<'b>(&'b mut self, crop_length: usize) -> &'b mut Query<'a> {
+        self.crop_length = Some(crop_length);
+        self
+    }
+    pub fn with_matches<'b>(&'b mut self, matches: bool) -> &'b mut Query<'a> {
+        self.matches = Some(matches);
+        self
+    }
+    pub fn build(&mut self) -> Query<'a> {
+        self.clone()
     }
```

Comment:
Maybe it would be better to consume the query itself `fn build(self) -> Query<'a> { self }` instead. This way we avoid cloning every time we query and if the user wants to do multiple queries with the same parameters he can clone the builder before building it.

Reply:
The problem here is that when we use the builder syntax, we return `&mut Self` at each method. So the build method, which is used at the end of the chain, cannot take `Self`. I think cloning is not a problem since it contains only options and references. Cloning is extremely cheap.
But what if we implement the `Future` trait and remove the build method?

Reply:
Yeah, you are right, this is a near free clone :)